src/play.jl

```diff
@@ -1,50 +1,132 @@
 """
-    GamePool
+    GamePool{N,S}
 
-A struct that defines a Wordle-like game.
+A struct that defines a Wordle-like game with targets of length `N`.
+`S <: Unsigned` is the smallest unsigned integer type that can store values in the
+range `0:(3 ^ N - 1)`, evaluated as `Wordlegames.scoretype(N)`
+
+The fields are:
+
+- `targetpool`: the target pool as a `Vector{NTuple{N,Char}}`
+- `guesspool`: a `Vector{NTuple{N,Char}}` of potential guesses (can be the same as `targets`)
+- `allscores`: a `Matrix{S}` where `allscores[i,j] = score(guesspool[j], targets[i])`
+- `active`: a `BitVector`. The active target pool is `targets[active]`.
+- `counts`: `Vector{Int}` of length `3 ^ N` in which bin counts are accumulated
+- `probs`: `Vector{Float64}` containing `counts ./ sum(counts)`
+- `guesses`: `Vector{NTuple{N,Char}}`
+- `guessinds`: `Vector{Int}` where the indexes into `guesspool` of guesses are stored
+- `scores`: scores of the guesses as a `Vector{S}`
+- `poolsizes`: `Vector{Int}` target pool size for prior to each guess being evaluated
+- `expected`: `Vector{Float32}` of expected pool sizes after each guess is scored
+- `entropy`: `Vector{Float32}` the (base2) entropy for each guess
+- `guesstype`: `Symbol` - either `:expected`, the default, or `:entropy`
+- `hardmode`: `Bool` - should the game be played in "Hard Mode"?
+
+Constructor signatures include
+
+    GamePool(targets::Vector{NTuple{N,Char}}, guesses::Vector{NTuple{N,Char}}; guesstype::Symbol=:expected, hardmode::Bool=true) where {N}
+    GamePool(pool::Vector{NTuple{N,Char}}; guesstype::Symbol=:expected, hardmode::Bool=true) where {N}
+    GamePool(pool::AbstractVector{<:AbstractString}; guesstype::Symbol=:expected, hardmode::Bool=true)
+    GamePool(pool::AbstractVector; guesstype::Symbol=:expected, hardmode::Bool=true)
 """
-struct GamePool
-    pool::Vector
-    allscores::Matrix{<:Unsigned}
-    firstguess::Int
-    firstexpectedpoolsize::Float64
-    active::Vector{Bool}
+struct GamePool{N,S}
+    targetpool::Vector{NTuple{N,Char}}
+    guesspool::Vector{NTuple{N,Char}}
+    allscores::Matrix{S}
+    active::BitVector
     counts::Vector{Int}
-    guess::Ref{Int}
-    expectedpoolsize::Ref{Float64}
-    lastscore::Ref{Int}
+    probs::Vector{Float32}
```

Comment:
This is Float64 in the docstring.

I would probably use Float64 for all of these things -- at most you're looking at 3 * 32 = 96 bytes of waste and Float32 will often not be more efficient on 64 bit architectures (due to alignment and register design). If we were going for GPUs, then that would be a different story. :smile:

Reply:
You're right.  I think I may delete the `probs` field and make it a property which can be calculated on the fly, using your next suggestion to write `entropybase2` with
```jl
    countsum = sum(counts)
    return sum(x -> xlogx(x / countsum) / logtwo, counts)
```